Revising the index page of the plugins section in our documentation (#12802)
conda/conda: docs/source/dev-guide/plugins/index.rst

========
Plugins
========

As of version ``22.11.0``, ``conda`` has support for user plugins, enabling extension and/or
alterations to some of its functionality.

Quick start
-----------

This is an example of a minimal working conda plugin that defines a new subcommand:

.. code-block:: python
   :caption: example_plugin.py

   import conda.plugins
   from conda.base.context import context

   def command(arguments: list[str]):
       print("Conda subcommand!")

   @conda.plugins.hookimpl
   def conda_subcommands():
       yield plugins.CondaSubcommand(

>            name="example", action=command, summary="Example of a conda subcommand"

Reviewer: the trialing comma trick!

```suggestion
           name="example",
           action=command,
           summary="Example of a conda subcommand",
```

   build-backend = "setuptools.build_meta"

   [project]
   name = "example"
   version = "1.0.0"
   description = "Example conda plugin"
   requires-python = ">=3.8"
   dependencies = ["conda"]

   [project.entry-points."conda"]
   example = "example"

.. code-block:: python
   :caption: setup.py

   from setuptools import setup

   setup(
       name="example",
       install_requires="conda",
       entry_points={"conda": ["example = example"]},

>        py_modules=["my_plugin"],

Author: ```suggestion
       py_modules=["example_plugin"],
```

Our package only consists a single Python module called ``example``. If you have a large project,
be sure to always point the entry point to the module containing the plugin hook declarations (i.e.
where ``conda.plugins.hookimpl`` is used).

A note on licensing
-------------------

When licensing plugins, we recommend using licenses such as BSD-3_, MIT_, and
`Apache License 2.0`_. Some ``import`` statements may possibly require the GPLv3_
license, which ensures that the software being licensed is open source.

Ultimately, the authors of the plugins can decide which license is best for their particular
use case. Be sure to credit the original author of the plugin, and keep in mind that
licenses can be altered depending on the situation.

For more information on which license to use for your custom plugin, please reference
the `"Choose an Open Source License"`_ site.
More examples
-------------

To see more examples of conda plugins, please visit the following resources:

API reference
-------------
- `conda-plugins-template`_: this is a repository with full examples that could be used a starting point for your plugin

.. autoattribute:: conda.plugins::hookimpl

>    :no-value:

Author: Yeah, that's a far ask.

Otherwise, it just looked really weird having this there:

<img width="385" alt="Bildschirmfoto 2023-06-20 um 12 13 44" src="https://github.com/conda/conda/assets/718726/f785e5f5-034e-463d-97a5-e2b5fa7cfc56">

I'm not sure that would really make sense for a new reader.

>    :no-value:

Reviewer 2: Sure, let's fix the docstring of the hookimpl

>    :no-value:

Author: Even if the doc string was better. I still do not think it would be a good fit for this particular page. The quick start already shows how to use it. I think this would be a better fit for a dedicated "API" page where users could go for a further detailed reference.

Using other plugin hooks
------------------------

   Conda plugin hook implementation marker.
For examples of how to use other plugin hooks, please read their respective documentation pages:

.. toctree::
   :maxdepth: 1

>    manager

Author: I removed it because it's not a plugin hook and isn't important for plugin authors to be aware of. The plugin manager should only ever be used inside of conda.

>    manager

Author: We now have the full Plugin API linked to at the bottom and this pull request has made some updates to this documentation too.

>    manager

Reviewer: so `manager.rst` should be deleted then?

also the `plugins-api` link added to the bottom doesn't appear anywhere in this page

>    manager

Reviewer 3: I removed that link at the bottom of the page (there's already a `:doc:`-style link in the API section)

> license, which ensures that the software being licensed is open source.

Author: 😬. Sounds like a good idea.

   build-backend = "setuptools.build_meta"

   [project]
   name = "conda-example-plugin"
   version = "1.0.0"
   description = "Example conda plugin"
   requires-python = ">=3.8"
   dependencies = ["conda"]

   [project.entry-points."conda"]
   conda-example-plugin = "example_plugin"

.. code-block:: python
   :caption: setup.py

   from setuptools import setup

   setup(
       name="conda-example-plugin",
       install_requires="conda",
       entry_points={"conda": ["conda-example-plugin = example_plugin"]},
       py_modules=["example_plugin"],
   )

In both examples shown above, we define an entry point for conda. It's important to make sure

> that the entry point is for "conda" and that it points to the correct module in your plugin package.

Reviewer 4: Would we want to put "conda" in backticks here?

Reviewer 3: Since this section is talking about entry points and we're referring to conda as the specific example, I think the quote marks are appropriate.

Reviewer 4: Sorry, just saw this. I meant doing `"conda"`, to be clear. But no worries either way.

Our package only consists a single Python module called ``example_plugin``. If you have a large project,
be sure to always point the entry point to the module containing the plugin hook declarations (i.e.
where ``conda.plugins.hookimpl`` is used). We recommend using the `plugin` submodule
in these cases, e.g. `large_project.plugin` (in `large_project/plugin.py`).

More examples
-------------

To see more examples of conda plugins, please visit the following resources:

> - `conda-plugins-template`_: this is a repository with full examples that could be used a starting point for your plugin

Reviewer 4: ```suggestion
- `conda-plugins-template`_: This is a repository with full examples that could be used a starting point for your plugin
```

Using other plugin hooks
------------------------

For examples of how to use other plugin hooks, please read their respective documentation pages:

.. toctree::
   :maxdepth: 1

   post_commands
   pre_commands
   solvers
   subcommands
   virtual_packages

More information about how plugins work
---------------------------------------

Plugins in ``conda`` are implemented with the use of Pluggy_, a Python framework used by
other projects, such as ``pytest``, ``tox``, and ``devpi``. ``pluggy`` provides the ability to
extend and modify the behavior of ``conda`` via function hooking, which results in plugin
systems that are discoverable with the use of `Python package entrypoints`_.

For more information about how it
works, we suggest heading over to their `documentation`_.

API
---

For even more detailed information about our plugin system, please the see the
:doc:`Plugin API </dev-guide/api/conda/plugins/index>` section.

A note on licensing
-------------------

For more information on which license to use for your custom plugin, please reference
the `"Choose an Open Source License"`_ site. If you need help figuring out exactly

> which one to use, we advise to seeking a qualified legal professional.

Reviewer 4: ```suggestion
which one to use, we advise communicating with a qualified legal professional.
```
Could also use "seeking out", but wasn't sure how colloquial that phrase was.

.. _`Apache License 2.0`: https://www.apache.org/licenses/LICENSE-2.0
.. _GPLv3: https://www.gnu.org/licenses/gpl-3.0.en.html
.. _`"Choose an Open Source License"`: https://choosealicense.com/

> .. _`conda-plugins-template`: https://github.com/conda/conda-plugin-template/tree/main/tutorials

Reviewer: ```suggestion
.. _`conda-plugins-template`: https://github.com/conda/conda-plugin-template
```
